homeassistant/components/rflink.py

```diff
@@ -0,0 +1,269 @@
+
+"""
+import asyncio
+import logging
+
+from homeassistant.const import (
+    ATTR_ENTITY_ID, EVENT_HOMEASSISTANT_STOP, STATE_UNKNOWN)
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['rflink==0.0.18']
+
+DOMAIN = 'rflink'
+
+RFLINK_EVENT = {
+    'light': 'rflink_switch_event_received',
+    'sensor': 'rflink_sensor_event_received',
+    'switch': 'rflink_switch_event_received',
+    'send_command': 'rflink_send_command',
+}
+
+ATTR_EVENT = 'event'
+ATTR_COMMAND = 'command'
+
+_LOGGER = logging.getLogger(__name__)
+
+KNOWN_DEVICE_IDS = []
+
+
+def identify_event_type(event):
+    """Look at event to determine type of device."""
+    if 'command' in event:
+        return 'light'
+    elif 'sensor' in event:
+        return 'sensor'
+    else:
+        return 'unknown'
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup the Rflink component."""
+    from rflink.protocol import create_rflink_connection
+
+    # make sure no known devices are left (mostly during tests)
+    del KNOWN_DEVICE_IDS[:]
+
+    def event_callback(event):
+        """Handle incoming rflink events.
+
+        Rflink events arrive as dictionaries of varying content
+        depending on their type. Identify the events and distribute
+        accordingly.
+
+        """
+        event_type = identify_event_type(event)
+        _LOGGER.info('event type %s', event_type)
+
+        # fire bus event for event type
+        if event_type in RFLINK_EVENT:
+            hass.bus.fire(RFLINK_EVENT[event_type], {ATTR_EVENT: event})
+        else:
+            _LOGGER.debug('unhandled event of type: %s', event_type)
+
+    # when connecting to tcp host instead of serial port (optional)
+    host = config[DOMAIN].get('host', None)
+    # tcp port when host configured, otherwise serial port
+    port = config[DOMAIN]['port']
+
+    # initiate serial/tcp connection to Rflink gateway
+    connection = create_rflink_connection(
+        port=port,
+        host=host,
+        event_callback=event_callback,
+        loop=hass.loop,
+        ignore=config[DOMAIN].get('ignore_devices', []),
+    )
+    transport, protocol = yield from connection
+    # handle shutdown of rflink asyncio transport
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                               lambda x: transport.close())
+
+    # provide channel for sending commands to rflink gateway
+    @asyncio.coroutine
+    def send_command_ack(event):
+        """Send command to rflink gateway via asyncio transport/protocol.
+
+        Waits for Rflink to confirm the packet has been sent.
+
+        """
+        yield from protocol.send_command_ack(
+            event.data[ATTR_ENTITY_ID],
+            event.data[ATTR_COMMAND],
+        )
+
+    def send_command(event):
+        """Send command to rflink gateway via asyncio transport/protocol.
+
+        Does not wait for Rflink to confirm the packet has been sent.
+
+        """
+        protocol.send_command(
```

Comment:
No, just a normal function, puts the command on the out buffer and returns, transport/protocol takes care of asynchronous write to the IO. Annotated this in the docstring for the function.